src/net_processing.cpp

```diff
@@ -2522,17 +2571,28 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
     const CBlockIndex *pindexLast = nullptr;
 
     // Do these headers connect to something in our block index?
-    bool headers_connect_blockindex{WITH_LOCK(::cs_main, return m_chainman.m_blockman.LookupBlockIndex(headers[0].hashPrevBlock) != nullptr)};
-    if (!headers_connect_blockindex) {
-        if (nCount <= MAX_BLOCKS_TO_ANNOUNCE) {
-            // If this looks like it could be a BIP 130 block announcement, use
-            // special logic for handling headers that don't connect, as this
-            // could be benign.
-            HandleFewUnconnectingHeaders(pfrom, peer, headers);
-        } else {
-            Misbehaving(peer, 10, "invalid header received");
+    {
+        LOCK(cs_main);
+        if (!LookupBlockIndexForPeer(pfrom, headers.front().hashPrevBlock)) {
+            CNodeState* node_state{State(pfrom.GetId())};
+            LogPrint(BCLog::NET, "received header %s: %s: %s (peer=%d, nUnconnectingHeaders=%d, nCount=%d)\n",
+                     headers[0].GetHash().ToString(),
+                     m_chainman.m_blockman.LookupBlockIndex(headers.front().hashPrevBlock) ?
+                         "we should not leak the existence of the previous block" :
+                         "missing prev block",
+                     headers[0].hashPrevBlock.ToString(),
+                     pfrom.GetId(), node_state->nUnconnectingHeaders,
+                     nCount);
+            if (nCount <= MAX_BLOCKS_TO_ANNOUNCE) {
+                // If this looks like it could be a BIP 130 block announcement, use
+                // special logic for handling headers that don't connect, as this
+                // could be benign.
+                HandleFewUnconnectingHeaders(pfrom, peer, headers);
+            } else {
+                Misbehaving(peer, 10, "invalid header received");
+            }
+            return;
         }
-        return;
     }
 
     // At this point, the headers connect to something in our block index.
```

Comment:
This comment is kinda outdated now (should add `for a given peer/network` or something?)?

Reply:
I think this comment is still accurate, at that point the headers will never not connect to something in the index.